### ISSUE TITLE
Checkpoint validation as an option

### ISSUE THREAD

**Issue by ChenchaoZhao** [huggingface/transformers#32067]
### Feature request

Added `validate_checkpoint_key: bool` in `from_pretrained`

### Motivation

When pytorch loads state dict there is an option `strict=True` but there is no counterpart in `from_pretrained`. If people saved the pretrained model incorrectly, the incorrect keys can silently pass through.

### Your contribution

There is already an validation step in the `from_pretrained` one just need to raise the error if `validate_checkpoint_key=True`

**Comment by amyeroberts** [comment]
Hi @ChenchaoZhao, thanks for opening a feature request! 

We do validation already within `from_pretrained` when we load in checkpoints, and raise errors if there's missing or unexpected keys in the state dict versus the model architecture. 

Perhaps you could expand a bit more on the kind of validation you'd like to see? 

> If people saved the pretrained model incorrectly, the incorrect keys can silently pass through.

Could you expand on this a bit, specifically how this might happen? Maybe a code snippet would help, as there could be a bug in the `save_pretrained` method

**Comment by ChenchaoZhao** [comment]
For example, my friend saved a model from distributed training without unwrapping. The state dict keys got a prefix that did not match the unwrapped model state dict keys. When loading the model `AutoModelForMaskedLM.from_pretrained` there was a warning saying every key was not matched. I hope the warning can be turned into a raise by a configurable kwarg in the `from_pretrained`

**Comment by amyeroberts** [comment]
@ChenchaoZhao OK, just to make sure I've correctly understood, what you're proposing is a flag which enforces strict loading such that if any weights in the model are not present in the state_dict then an exception is raised?

**Comment by ChenchaoZhao** [comment]
> @ChenchaoZhao OK, just to make sure I've correctly understood, what you're proposing is a flag which enforces strict loading such that if any weights in the model are not present in the state_dict then an exception is raised? 

That's correct. Thanks!

**Comment by ArthurZucker** [comment]
Yeah sounds good. In general people load base model inside specifc §`ForZZZ` classes, which is why we don't have this by default. Not sure how much usage this will get but feel free to open a PR, we can make sure this is used in every single one of our "conversion scripts" to ensure we don't make mistakes! (happened to me!)

**Comment by qgallouedec** [comment]
Hey @ChenchaoZhao, sorry for the slow reply, catching up on the backlog. Arthur's +1 above still stands, but nobody has picked this up since, and `from_pretrained` still has no strict-loading flag today: `output_loading_info=True` hands you the missing/unexpected key sets, but nothing raises on them automatically.

That's roughly where a `strict` kwarg would hook in, right after `loading_info` is finalized. Leaving this open

**Comment by YeonwooSung** [comment]
@qgallouedec @ArthurZucker checking before I open anything.

`from_pretrained` still only has `output_loading_info=True` and no way to raise on a key mismatch. I would keep today's default and add an opt-in (`strict=True` or similar) so existing `For*` loads stay unchanged.

Two questions:
1. Is an opt-in flag still the shape you want?
2. Should it raise on missing keys, unexpected keys, or both?

If that is fine I can send a small PR with a regression test. I will wait for an ACK.

**Comment by YeonwooSung** [comment]
@qgallouedec @ArthurZucker No ACK on the two questions after 8 days, so I am going with:

- opt-in `strict=True`
- raise on missing or unexpected keys (same contract as `torch.nn.Module.load_state_dict(strict=True)`)
- default stays `False`, so `For*` loads of a base checkpoint stay unchanged

Opening a small PR with a regression test. If you wanted unexpected-only, I can tighten it.
